### PR TITLE
Vault: Fix reward fee issue 

### PR DESCRIPTION
**Pull request by coachchucksol**
A simpler solution to:

https://github.com/jito-foundation/restaking/pull/136

### CODE REVIEW

File: integration_tests/tests/vault/update_token_metadata.rs [hunk 1]
```diff
@@ -1,124 +1,191 @@
 #[cfg(test)]
```
Comment (ebatsell) [inline integration_tests/tests/vault/update_token_metadata.rs:1]: I think some of these tests are mixing up update_vault_balance with update_token_metadata?

Comment (coachchucksol) [reply integration_tests/tests/vault/update_token_metadata.rs:1]: Oh boo, you're correct. Thanks for catching that.

Comment (coachchucksol) [reply integration_tests/tests/vault/update_token_metadata.rs]: fixed

File: vault_program/src/update_vault_balance.rs [hunk 2]
```diff
@@ -37,23 +37,34 @@ pub fn process_update_vault_balance(
...
+
+    // 3. Calculate the reward fee in VRT
+    let vrt_reward_fee = vault.calculate_vrt_mint_amount(st_reward_fee)?;
+
+    // 4. Update State, with the vrt fee and the new st balance
+    vault.set_tokens_deposited(new_st_balance);
+    vault.increment_vrt_supply(vrt_reward_fee)?;
+
+    // 5. Check for rewards not substantial enough
+    vault.check_reward_fee_effective_rate(st_rewards, vrt_reward_fee, 50)?;
```
Comment (ebatsell) [inline vault_program/src/update_vault_balance.rs:57]: make 50 a constant

Comment (coachchucksol) [reply vault_program/src/update_vault_balance.rs:57]: good call

**Review by ebatsell — APPROVED** [review]

File: vault_program/src/update_vault_balance.rs [hunk 3]
```diff
@@ -37,23 +37,38 @@ pub fn process_update_vault_balance(
...
-    let reward_fee = vault.calculate_rewards_fee(new_balance)?;
+    // 1. Calculate reward fee in VRT
+    let st_rewards = new_st_balance.saturating_sub(vault.tokens_deposited());
+    let st_reward_fee = vault.calculate_st_reward_fee(new_st_balance)?;
 
-    // Update state
-    vault.set_tokens_deposited(new_balance);
-    vault.increment_vrt_supply(reward_fee)?;
+    // 2. Increment ST less the reward fee
+    vault.set_tokens_deposited(new_st_balance.checked_sub(st_reward_fee).unwrap());
```
Comment (buffalu) [inline vault_program/src/update_vault_balance.rs:47]: nit: unwrap

File: vault_program/src/update_vault_balance.rs [hunk 4]
```diff
@@ -37,23 +37,38 @@ pub fn process_update_vault_balance(
     vault.check_is_paused()?;
 
     // Calculate rewards
-    let new_balance = Account::unpack(&vault_token_account.data.borrow())?.amount;
+    let new_st_balance = Account::unpack(&vault_token_account.data.borrow())?.amount;
 
-    let reward_fee = vault.calculate_rewards_fee(new_balance)?;
+    // 1. Calculate reward fee in VRT
```
Comment (buffalu) [inline vault_program/src/update_vault_balance.rs:42]: this is in ST?

File: vault_program/src/update_vault_balance.rs (hunk 3 above)
Comment (buffalu) [reply vault_program/src/update_vault_balance.rs:47]: also im kinda curious what the checked_sub accomplishes compared to just setting new_st_balance?

**Review by buffalu — APPROVED** [review]

File: vault_program/src/update_vault_balance.rs (hunk 3 above)
Comment (coachchucksol) [reply vault_program/src/update_vault_balance.rs:47]: What's happening:

1. We take our fee in st
2. We add the reward ( total reward - fee in st )
3. We virtually call `mint_to` on the reward fee ob behalf of the vault

File: vault_program/src/update_vault_balance.rs (hunk 4 above)
Comment (coachchucksol) [reply vault_program/src/update_vault_balance.rs:42]: you're correct, it's ST

**Pull request merged**